File: apps/multipass/platforms/windows/PairingDialogs.cs
namespace Multipass;

/// <summary>Shows the six-digit code for the pairing in progress. The engine owns the pairing.</summary>
internal sealed class PairingDialog : Form
{
    private readonly Action<bool> decide;
    private bool settled;

    public PairingDialog(string peerName, string? code, bool incoming, Action<bool> decide)
    {
        this.decide = decide;
        Text = "Pair with another computer";
        FormBorderStyle = FormBorderStyle.FixedDialog;
        StartPosition = FormStartPosition.CenterParent;
        MinimizeBox = MaximizeBox = false;
        ShowInTaskbar = false;
        ClientSize = new Size(480, 230);
        var layout = new FlowLayoutPanel { Dock = DockStyle.Fill, Padding = new Padding(18), FlowDirection = FlowDirection.TopDown, WrapContents = false };
        var buttons = new FlowLayoutPanel { AutoSize = true, WrapContents = false, FlowDirection = FlowDirection.LeftToRight, Margin = new Padding(3, 12, 3, 3) };
        if (code is null)
        {
            layout.Controls.Add(new Label { Text = $"Connecting to “{peerName}”…", AutoSize = true, MaximumSize = new Size(440, 0) });
            var cancel = new Button { Text = "Cancel", AutoSize = true };
            cancel.Click += (_, _) => Answer(false);
            buttons.Controls.Add(cancel);
        }
        else
        {
            var heading = incoming ? $"“{peerName}” wants to pair with this computer." : $"Pairing with “{peerName}”.";
            layout.Controls.Add(new Label { Text = heading, AutoSize = true, MaximumSize = new Size(440, 0), Font = new Font(Font, FontStyle.Bold) });
            layout.Controls.Add(new Label
            {
                Text = Spaced(code), AutoSize = true, Margin = new Padding(3, 10, 3, 10),
                Font = new Font(FontFamily.GenericMonospace, 28, FontStyle.Bold),
            });
            layout.Controls.Add(new Label { Text = $"Confirm only if {peerName} shows the same code. Both computers must confirm.", AutoSize = true, MaximumSize = new Size(440, 0) });
            var decline = new Button { Text = incoming ? "Decline" : "Cancel", AutoSize = true };
            var accept = new Button { Text = incoming ? "Accept" : "Confirm", AutoSize = true };
            decline.Click += (_, _) => Answer(false);
            accept.Click += (_, _) => Answer(true);
            buttons.Controls.AddRange([decline, accept]);
            AcceptButton = accept;
        }
        layout.Controls.Add(buttons);
        Controls.Add(layout);
        FormClosing += (_, _) =>
        {
            // Closing with the title-bar button declines, unless the engine already finished.
            if (!settled) { settled = true; decide(false); }
        };
    }

    private static string Spaced(string code) => code.Length == 6 ? $"{code[..3]} {code[3..]}" : code;

    private void Answer(bool accept)
    {
        if (settled) return;
        settled = true;
        Close();
        decide(accept);
    }

    /// <summary>Close because the engine reported the pairing finished; sends no decision.</summary>
    public void Settle()
    {
        settled = true;
        Close();
    }
}

File: apps/multipass/platforms/linux/pairing_dialogs.py
"""Transient native pairing prompt. The engine owns the pairing; this only shows the code."""
from gi.repository import Gtk


def spaced(code):
    return f"{code[:3]} {code[3:]}" if len(code) == 6 else code


def show_pairing(parent, pairing, decide):
    """Show the code for the pairing in progress. Returns the window; close it with
    ``settle()`` when the engine reports the pairing finished, so closing does not
    count as a decline."""
    name = pairing["peer_name"]
    incoming = pairing["incoming"]
    code = pairing.get("code")
    window = Gtk.Window(title="Pair with another computer", transient_for=parent, modal=True, destroy_with_parent=True)
    window.set_default_size(460, 220)
    box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=12)
    for setter in (box.set_margin_top, box.set_margin_bottom, box.set_margin_start, box.set_margin_end):
        setter(20)
    window.set_child(box)
    buttons = Gtk.Box(spacing=12, halign=Gtk.Align.END)
    state = {"settled": False}

    def answer(accept):
        if state["settled"]:
            return
        state["settled"] = True
        window.close()
        decide(accept)

    if code is None:
        box.append(Gtk.Label(label=f"Connecting to “{name}”…", xalign=0, wrap=True))
        cancel = Gtk.Button(label="Cancel")
        cancel.connect("clicked", lambda _: answer(False))
        buttons.append(cancel)
    else:
        heading = f"“{name}” wants to pair with this computer." if incoming else f"Pairing with “{name}”."
        box.append(Gtk.Label(label=heading, xalign=0, wrap=True))
        digits = Gtk.Label(label=spaced(code), selectable=True)
        digits.add_css_class("title-1")
        box.append(digits)
        box.append(Gtk.Label(label=f"Confirm only if {name} shows the same code. Both computers must confirm.", xalign=0, wrap=True))
        decline = Gtk.Button(label="Decline" if incoming else "Cancel")
        accept = Gtk.Button(label="Accept" if incoming else "Confirm")
        accept.add_css_class("suggested-action")
        decline.connect("clicked", lambda _: answer(False))
        accept.connect("clicked", lambda _: answer(True))
        buttons.append(decline)
        buttons.append(accept)
        window.set_default_widget(accept)
    box.append(buttons)

    def closed(*_):
        # Closing the prompt with the window button declines, unless the engine already finished.
        if not state["settled"]:
            state["settled"] = True
            decide(False)
        return False

    def settle():
        state["settled"] = True
        window.close()

    window.settle = settle
    window.connect("close-request", closed)
    window.present()
    return window
